src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +226,163 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
             }
         }
 
-        private async Task PublishDiagnosticsAsync(Document document)
+        /// <summary>
+        /// Stores the last published LSP diagnostics with the Roslyn document that they came from.
+        /// This is useful in the following scenario.  Imagine we have documentA which has contributions from mapped files m1 and m2.
+        /// dA -> m1
+        /// And m1 additionally contributes to documentB.
+        /// m1 -> dA, dB
+        /// When we query for diagnostic on dA, we get a subset of the diagnostics on m1 (missing the contributions from dB)
+        /// Since each publish diagnostics notification replaces diagnostics per document,
+        /// we must union the diagnostics contribution from dB and dA to produce all diagnostics for m1 and publish all at once.
+        /// 
+        /// This dictionary stores the previously computed diagnostics for the published file so that we can
+        /// union the currently computed diagnostics (e.g. for dA) with previously computed diagnostics (e.g. from dB).
+        /// </summary>
+        private readonly Dictionary<Uri, Dictionary<DocumentId, ImmutableArray<LanguageServer.Protocol.Diagnostic>>> _publishedFileToDiagnostics =
+            new Dictionary<Uri, Dictionary<DocumentId, ImmutableArray<LanguageServer.Protocol.Diagnostic>>>();
+
+        /// <summary>
+        /// Stores the mapping of a document to the uri(s) of diagnostics previously produced for this document.
+        /// When we get empty diagnostics for the document we need to find the uris we previously published for this document.
+        /// Then we can publish the updated diagnostics set for those uris (either empty or the diagnostic contributions from other documents).
+        /// </summary>
+        private readonly Dictionary<DocumentId, ImmutableHashSet<Uri>> _documentsToPublishedUris = new Dictionary<DocumentId, ImmutableHashSet<Uri>>();
+
+        internal async Task PublishDiagnosticsAsync(Document document)
```

Comment:
Is this serialized so we don't have to worry about parallel operations?

Reply:
Yes, and so that we publish diagnostics 'in order' - since diagnostics are replaced we don't want to publish old diagnostics after new diagnostics.